src/PlatformBenchmarks/HttpApplication.cs

```diff
@@ -64,85 +67,90 @@ public virtual ValueTask OnReadCompletedAsync()
-                        await OnReadCompletedAsync();
-                    }
+                var result = await task;
+                var buffer = result.Buffer;
+                var consumed = buffer.Start;
+                var examined = buffer.End;
 
-                    var result = await task;
-                    var buffer = result.Buffer;
-                    var consumed = buffer.Start;
-                    var examined = buffer.End;
+                if (!buffer.IsEmpty)
+                {
+                    ParseHttpRequest(buffer, out consumed, out examined);
 
-                    try
-                    {
-                        if (!buffer.IsEmpty)
-                        {
-                            ParseHttpRequest(buffer, out consumed, out examined);
-
-                            if (_state != State.Body && result.IsCompleted)
-                            {
-                                throw new InvalidOperationException("Unexpected end of data!");
-                            }
-                        }
-                        else if (result.IsCompleted)
-                        {
-                            break;
-                        }
-                    }
-                    finally
+                    if (_state != State.Body && result.IsCompleted)
                     {
-                        Connection.Transport.Input.AdvanceTo(consumed, examined);
+                        ThrowUnexpectedEndOfData();
                     }
+                }
+                else if (result.IsCompleted)
+                {
+                    break;
+                }
 
-                    if (_state == State.Body)
-                    {
-                        await ProcessRequestAsync();
+                Reader.AdvanceTo(consumed, examined);
 
-                        _state = State.StartLine;
-                    }
-                }
+                if (_state == State.Body)
+                {
+                    await ProcessRequestAsync();
 
-                Connection.Transport.Input.Complete();
-            }
-            catch (Exception ex)
-            {
-                Connection.Transport.Input.Complete(ex);
-            }
-            finally
-            {
-                Connection.Transport.Output.Complete();
+                    _state = State.StartLine;
+                }
             }
         }
 
-        private void ParseHttpRequest(ReadOnlySequence<byte> buffer, out SequencePosition consumed, out SequencePosition examined)
+        private void ParseHttpRequest(in ReadOnlySequence<byte> buffer, out SequencePosition consumed, out SequencePosition examined)
         {
             consumed = buffer.Start;
             examined = buffer.End;
 
-            if (_state == State.StartLine)
+            var parsingStartLine = _state == State.StartLine;
+            if (parsingStartLine)
             {
                 if (Parser.ParseRequestLine(this, buffer, out consumed, out examined))
```

Comment:
Would it make sense to allow passing buffer by-in?

Reply:
Yes, we just changed this in Kestrel.